baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+"""
+==============================================================================
+BaseClasses: Solver History Class
+==============================================================================
+
+@Description : A general solver history class for storing values from a nonlinear solution.
+"""
+
+# ==============================================================================
+# Standard Python modules
+# ==============================================================================
+from collections import OrderedDict
+import copy
+import os
+import time
+from typing import Optional, Type, Dict, Any, List, Iterable, Union
+import pickle
+
+
+class SolverHistory(object):
+    """The SolverHistory class can be used to store and print various useful values during the execution of a solver.
+
+    NOTE: The implementation of this class contains no consideration of parallelism. If you are using a solverHistory
+    object in your parallel solver, you will need to take care over which procs you make calls to the SolverHistory
+    object on.
+    """
+
+    __slots__ = [
+        "variables",
+        "data",
+        "metadata",
+        "iter",
+        "startTime",
+        "defaultFormat",
+        "testValues",
+        "DEFAULT_OTHER_FORMAT",
+        "DEFAULT_OTHER_VALUE",
+    ]
+
+    def __init__(self) -> None:
+        # Dictionaries for storing variable information and values
+        self.variables: Dict[str, Any] = OrderedDict()
+        self.data: Dict[str, List] = OrderedDict()
+        self.metadata: Dict[str, Any] = {}
+
+        # Initialise iteration counter and solve start time
+        self.iter: int = 0
+        self.startTime: float = -1.0
+
+        # --- Define default print formatting for some common types ---
+        self.defaultFormat: Dict[Type, str] = {}
+        # Test values are used to check how wide the printed formatted values of a variable will be, so that we can make
+        # sure the columns in the iteration printout are wide enough
+        self.testValues: Dict[Type, Any] = {}
+        # float
+        self.defaultFormat[float] = "{: 17.11e}"
+        self.testValues[float] = 0.0
+        # int
+        self.defaultFormat[int] = "{: 5d}"
+        self.testValues[int] = 0
+        # str
+        self.defaultFormat[str] = "{:^10}"
+        self.testValues[str] = "a"
+        # other
```

Comment:
Maybe add default format for complex numbers?

Reply:
Done